README.md
# Scala Exercises - Monocle

This repository hosts a template content library for the [Scala Exercises](https://www.scala-exercises.org/) platform, including interactive exercises related to the [Monocle](https://github.com/julien-truffaut/Monocle) optics library for Scala (and [Scala.js](https://www.scala-js.org/)).

------------------------


## Run Locally

- Clone this repository, compile and publish the project:

~~~ sh
git clone git@github.com:scala-exercises/exercises-monocle.git
cd exercises-monocle/
sbt compile publishLocal # it is important to first run the `compile` command alone
~~~

- Clone the `evaluator` and run it:

~~~ sh
git clone git@github.com:scala-exercises/evaluator.git
cd evaluator/
sbt "project evaluator-server" run
~~~

- Clone the `scala-exercises` fork:

~~~ sh
git clone git@github.com:scala-exercises/scala-exercises.git
~~~

- Follow the database setup instructions given
  [here](https://github.com/scala-exercises/scala-exercises#configure-the-database)

- Add the following line the `server/conf/application.dev.conf`:

~~~
evaluator.secretKey="secretKey"
~~~

- Run the server:

~~~ sh
sbt -mem 1500 run
~~~

## About Scala exercises

"Scala Exercises" brings exercises for the Stdlib, Cats, Shapeless and many other great libraries for Scala to your browser. Offering hundreds of solvable exercises organized into several categories covering the basics of the Scala language and it's most important libraries.

Scala Exercises is available at [scala-exercises.org](https://scala-exercises.org).

## Contributing

Contributions welcome! Please join our [Gitter channel](https://gitter.im/scala-exercises/scala-exercises)
to get involved, or visit our [GitHub site](https://github.com/scala-exercises).

## License

Copyright (C) 2015-2016 47 Degrees, LLC.
Reactive, scalable software solutions.
http://47deg.com
hello@47deg.com

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

## [v0.6.6](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.6) (2020-08-10)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/v0.6.5...v0.6.6)

📈 **Dependency updates**

- Update definitions, exercise-compiler, ... to 0.6.4 [\#54](https://github.com/scala-exercises/exercises-monocle/pull/54) ([47erbot](https://github.com/47erbot))

## [v0.6.5](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.5) (2020-08-10)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/v0.6.4...v0.6.5)

📈 **Dependency updates**

- Update sbt-mdoc to 2.2.4 [\#53](https://github.com/scala-exercises/exercises-monocle/pull/53) ([47erbot](https://github.com/47erbot))
- Update scalacheck-1-14 to 3.2.1.0 [\#52](https://github.com/scala-exercises/exercises-monocle/pull/52) ([47erbot](https://github.com/47erbot))
- Update scalatest to 3.2.1 [\#51](https://github.com/scala-exercises/exercises-monocle/pull/51) ([47erbot](https://github.com/47erbot))
- Update sbt-scalafmt to 2.4.2 [\#49](https://github.com/scala-exercises/exercises-monocle/pull/49) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.4 [\#48](https://github.com/scala-exercises/exercises-monocle/pull/48) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.3 [\#47](https://github.com/scala-exercises/exercises-monocle/pull/47) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.2 [\#45](https://github.com/scala-exercises/exercises-monocle/pull/45) ([scala-steward](https://github.com/scala-steward))

## [v0.6.4](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.4) (2020-06-30)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/v0.6.3...v0.6.4)

## [v0.6.3](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.3) (2020-06-29)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/v0.6.2...v0.6.3)

📈 **Dependency updates**

- Update sbt to 1.3.13 [\#42](https://github.com/scala-exercises/exercises-monocle/pull/42) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.3 [\#41](https://github.com/scala-exercises/exercises-monocle/pull/41) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-1-14 to 3.2.0.0 [\#40](https://github.com/scala-exercises/exercises-monocle/pull/40) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.2.0 [\#39](https://github.com/scala-exercises/exercises-monocle/pull/39) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.1 [\#38](https://github.com/scala-exercises/exercises-monocle/pull/38) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.2 [\#37](https://github.com/scala-exercises/exercises-monocle/pull/37) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.11 [\#35](https://github.com/scala-exercises/exercises-monocle/pull/35) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.1.2 [\#34](https://github.com/scala-exercises/exercises-monocle/pull/34) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Update definitions, exercise-compiler, ... to 0.6.3 [\#43](https://github.com/scala-exercises/exercises-monocle/pull/43) ([scala-steward](https://github.com/scala-steward))
- Prepare repository for next  release and SBT build improvements [\#36](https://github.com/scala-exercises/exercises-monocle/pull/36) ([juanpedromoreno](https://github.com/juanpedromoreno))

## [v0.6.2](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.2) (2020-04-27)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/v0.6.1...v0.6.2)

## [v0.6.1](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.1) (2020-04-27)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/v0.6.0...v0.6.1)

**Merged pull requests:**

- Compile when Publish [\#31](https://github.com/scala-exercises/exercises-monocle/pull/31) ([juanpedromoreno](https://github.com/juanpedromoreno))

## [v0.6.0](https://github.com/scala-exercises/exercises-monocle/tree/v0.6.0) (2020-04-24)

[Full Changelog](https://github.com/scala-exercises/exercises-monocle/compare/dc174cab40ee92cfc040c40ff8bb0575738f5952...v0.6.0)

**Closed issues:**

- Exercises with nothing to fill in / impossible to complete [\#1](https://github.com/scala-exercises/exercises-monocle/issues/1)

**Merged pull requests:**

- Updates Project [\#27](https://github.com/scala-exercises/exercises-monocle/pull/27) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Fix for markup and formatting [\#25](https://github.com/scala-exercises/exercises-monocle/pull/25) ([hobovsky](https://github.com/hobovsky))
- Update sbt to 1.3.10 [\#24](https://github.com/scala-exercises/exercises-monocle/pull/24) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.9 [\#23](https://github.com/scala-exercises/exercises-monocle/pull/23) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-shapeless\_1.14 to 1.2.5 [\#22](https://github.com/scala-exercises/exercises-monocle/pull/22) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.1.1 [\#21](https://github.com/scala-exercises/exercises-monocle/pull/21) ([scala-steward](https://github.com/scala-steward))
- Mergify: configuration update [\#20](https://github.com/scala-exercises/exercises-monocle/pull/20) ([angoglez](https://github.com/angoglez))
- Update scalacheck-shapeless\_1.14 to 1.2.4 [\#19](https://github.com/scala-exercises/exercises-monocle/pull/19) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.8 [\#18](https://github.com/scala-exercises/exercises-monocle/pull/18) ([scala-steward](https://github.com/scala-steward))

src/test/scala/monocle/IsoSpec.scala
/*
 * Copyright 2017-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monoclelib

import IsoHelper.Person
import org.scalacheck.ScalacheckShapeless._
import org.scalaexercises.Test
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.scalacheck.Checkers
import shapeless.HNil

class IsoSpec extends AnyFunSuite with Checkers {

  test("exercise person to tuple") {
    check(
      Test.testSuccess(
        IsoExercises.exercisePersonToTuple _,
        ("Zoe", 25) :: Person("Zoe", 25) :: HNil
      )
    )
  }

  test("exercise person to tuple apply") {
    check(
      Test.testSuccess(
        IsoExercises.exercisePersonToTupleApply _,
        Person("Zoe", 25) :: HNil
      )
    )
  }

  test("exercise list to vector") {
    check(
      Test.testSuccess(
        IsoExercises.exerciseListToVector _,
        Vector(1, 2, 3) :: HNil
      )
    )
  }

  test("exercise string to list") {
    check(
      Test.testSuccess(
        IsoExercises.exerciseStringToList _,
        "ello" :: HNil
      )
    )
  }

  test("exercise gen iso apply") {
    check(
      Test.testSuccess(
        IsoExercises.exerciseGenIsoApply _,
        "Hello" :: HNil
      )
    )
  }

  test("exercise gen iso fields") {
    check(
      Test.testSuccess(
        IsoExercises.exerciseGenIsoFields _,
        ("John", 42) :: HNil
      )
    )
  }

  test("exercise laws") {
    check(
      Test.testSuccess(
        IsoExercises.exerciseLaws _,
        true :: true :: HNil
      )
    )
  }
}


src/test/scala/monocle/LensSpec.scala
/*
 * Copyright 2017-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monoclelib

import LensHelper.{Address, Person, Point}
import org.scalacheck.{Arbitrary, Gen}
import org.scalacheck.ScalacheckShapeless._
import org.scalaexercises.Test
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.scalacheck.Checkers
import shapeless.HNil

class LensSpec extends AnyFunSuite with Checkers {

  implicit val addressArbitrary = Arbitrary {
    for {
      streetName   <- Gen.identifier
      streetNumber <- Gen.posNum[Int]
    } yield Address(streetNumber, streetName)
  }

  implicit val personArbitrary = Arbitrary {
    for {
      name    <- Gen.identifier
      age     <- Gen.posNum[Int]
      address <- addressArbitrary.arbitrary
    } yield Person(name, age, address)
  }

  test("exercise get and set") {
    check(
      Test.testSuccess(
        LensExercises.exerciseGetAndSet _,
        10 :: Address(5, "High Street") :: HNil
      )
    )
  }

  test("exercise modify") {
    check(
      Test.testSuccess(
        LensExercises.exerciseModify _,
        Address(11, "High Street") :: 10 :: Address(11, "High Street") :: HNil
      )
    )
  }

  test("exercise modifyF") {
    check(
      Test.testSuccess(
        LensExercises.exerciseModifyF _,
        List(Address(9, "High Street"), Address(11, "High Street")) :: List(
          Address(134, "High Street"),
          Address(136, "High Street")
        ) :: HNil
      )
    )
  }

  test("exercise compose lens") {
    check(
      Test.testSuccess(
        LensExercises.exerciseComposeLens _,
        10 :: Person("John", 20, Address(2, "High Street")) :: HNil
      )
    )
  }

  test("exercise compose macro") {
    check(
      Test.testSuccess(
        LensExercises.exerciseComposeMacro _,
        Person("Mike", 21, LensHelper.address) :: HNil
      )
    )
  }

  test("exercise clens prism composition") {
    check(Test.testSuccess(LensExercises.exerciseLensPrismComposition _, Option(1) :: HNil))
  }

  test("exercise lens generation") {
    check(
      Test.testSuccess(
        LensExercises.exerciseLensGeneration _,
        Person("John", 20, Address(10, "Iffley Road")) :: HNil
      )
    )
  }

  test("exercise lens macro annotation") {
    check(Test.testSuccess(LensExercises.exerciseLensMacroAnn _, 5 :: Point(5, 0) :: HNil))
  }

  test("exercise lens macro annotation prefix") {
    check(Test.testSuccess(LensExercises.exerciseLensMacroAnnPrefix _, 5 :: HNil))
  }

  test("exercise laws") {
    check(
      Test.testSuccess(
        LensExercises.exerciseLaws _,
        true :: true :: HNil
      )
    )
  }
}


src/test/scala/monocle/PrismSpec.scala
/*
 * Copyright 2017-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monoclelib

import PrismHelper.{JNum, JStr}
import org.scalaexercises.Test
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.scalacheck.Checkers

import org.scalacheck.ScalacheckShapeless._
import shapeless.HNil

class PrismSpec extends AnyFunSuite with Checkers {

  test("exercise get option apply") {
    check(
      Test.testSuccess(
        PrismExercises.exerciseGetOptionAndApply _,
        JStr("hello") :: Option("Hello") :: Option.empty[String] :: HNil
      )
    )
  }

  test("exercise set and modify") {
    check(
      Test.testSuccess(
        PrismExercises.exerciseSetAndModify _,
        JStr("Bar") :: JStr("olleH") :: HNil
      )
    )
  }

  test("exercise set and modify 2") {
    check(
      Test.testSuccess(
        PrismExercises.exerciseSetAndModify2 _,
        JNum(10) :: JNum(10) :: HNil
      )
    )
  }

  test("exercise modifyOption") {
    check(
      Test.testSuccess(
        PrismExercises.exerciseModifyOption _,
        Option(JStr("olleH")) :: Option.empty[JNum] :: HNil
      )
    )
  }

  test("exercise compose") {
    check(
      Test.testSuccess(
        PrismExercises.exerciseCompose _,
        JNum(5.0) :: Option(5) :: Option.empty[Int] :: Option.empty[String] :: HNil
      )
    )
  }

  test("exercise prism generation") {
    check(
      Test.testSuccess(
        PrismExercises.exercisePrismGeneration _,
        Option(JNum(4.5)) :: Option.empty[JNum] :: HNil
      )
    )
  }

  test("exercise laws") {
    check(
      Test.testSuccess(
        PrismExercises.exerciseLaws _,
        true :: true :: HNil
      )
    )
  }
}


src/main/scala/monocle/MonocleLib.scala
/*
 * Copyright 2017-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monoclelib

import org.scalaexercises.definitions._

/**
 * Monocle is an optics library for Scala (and Scala.js) strongly inspired by Haskell Lens.
 *
 * @param name monocle
 */
object MonocleLib extends Library {
  override def owner = "scala-exercises"

  override def repository = "exercises-monocle"

  override def color = Some("#5B5988")

  override def sections =
    List(
      IsoExercises,
      LensExercises,
      PrismExercises,
      OptionalExercises,
      TraversalExercises
    )
  override def logoPath = "monocle"
}


src/main/scala/monocle/IsoExercises.scala
/*
 * Copyright 2017-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monoclelib

import monocle.Iso
import monocle.macros.GenIso
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import org.scalaexercises.definitions._

object IsoHelper {
  case class Person(name: String, age: Int)

  val personToTuple = Iso[Person, (String, Int)](p => (p.name, p.age)) {
    case (name, age) => Person(name, age)
  }

  def listToVector[A] = Iso[List[A], Vector[A]](_.toVector)(_.toList)

  def vectorToList[A] = listToVector[A].reverse

  val stringToList = Iso[String, List[Char]](_.toList)(_.mkString(""))

  case class MyString(s: String)
  case class Foo()
  case object Bar

}

/**
 * == Iso ==
 *
 * An [[http://julien-truffaut.github.io/Monocle/optics/iso.html `Iso`]] is an optic which converts elements of type `S` into elements of type `A` without loss.
 *
 * Consider a case class `Person` with two fields:
 *
 * {{{
 *   case class Person(name: String, age: Int)
 * }}}
 *
 * @param name iso
 */
object IsoExercises extends AnyFlatSpec with Matchers with Section {

  import IsoHelper._

  /**
   * `Person` is equivalent to a tuple `(String, Int)` and a tuple `(String, Int)` is equivalent to `Person`. So we can create an `Iso` between `Person` and `(String, Int)` using two total functions:
   *
   *  - `get: Person => (String, Int)`
   *  - `reverseGet (aka apply): (String, Int) => Person`
   *
   * {{{
   *   import monocle.Iso
   *   val personToTuple = Iso[Person, (String, Int)](p => (p.name, p.age)){case (name, age) => Person(name, age)}
   * }}}
   */
  def exercisePersonToTuple(res0: (String, Int), res1: Person) = {

    personToTuple.get(Person("Zoe", 25)) should be(res0)
    personToTuple.reverseGet(("Zoe", 25)) should be(res1)
  }

  /**
   * Or simply:
   */
  def exercisePersonToTupleApply(res0: Person) =
    personToTuple(("Zoe", 25)) should be(res0)

  /**
   * Another common use of `Iso` is between collection. `List` and `Vector` represent the same concept, they are both an ordered sequence of elements but they have different performance characteristics. Therefore, we can define an `Iso` between a `List[A]` and a `Vector[A]`:
   * {{{
   *   def listToVector[A] = Iso[List[A], Vector[A]](_.toVector)(_.toList)
   * }}}
   */
  def exerciseListToVector(res0: Vector[Int]) =
    listToVector.get(List(1, 2, 3)) should be(res0)

  /**
   * We can also `reverse` an `Iso` since it defines a symmetric transformation:
   * {{{
   *   def vectorToList[A] = listToVector[A].reverse
   *     // vectorToList: [A]=> monocle.PIso[Vector[A],Vector[A],List[A],List[A]]
   * }}}
   */
  def exerciseVectorToList(res0: List[Int]) =
    vectorToList.get(Vector(1, 2, 3)) should be(res0)

  /**
   * `Iso` are also convenient to lift methods from one type to another, for example a `String` can be seen as a `List[Char]` so we should be able to transform all functions `List[Char] => List[Char]` into `String => String`:
   * {{{
   *   val stringToList = Iso[String, List[Char]](_.toList)(_.mkString(""))
   * }}}
   */
  def exerciseStringToList(res0: String) =
    stringToList.modify(_.tail)("Hello") should be(res0)

  /**
   * = Iso Generation =
   *
   * We defined several macros to simplify the generation of `Iso` between a case class and its `Tuple` equivalent. All macros are defined in a separate module (see modules).
   * {{{
   *     case class MyString(s: String)
   *     case class Foo()
   *     case object Bar
   *
   *     import monocle.macros.GenIso
   * }}}
   *
   * First of all, `GenIso.apply` generates an `Iso` for `newtype` i.e. case class with a single type parameter:
   */
  def exerciseGenIsoApply(res0: String) =
    GenIso[MyString, String].get(MyString("Hello")) should be(res0)

  /**
   * Then, `GenIso.unit` generates an `Iso` for object or case classes with no field:
   *
   * {{{
   * GenIso.unit[Foo]
   * // res8: monocle.Iso[Foo,Unit] = monocle.PIso$$anon$10@280a5b3b
   * GenIso.unit[Bar.type]
   * // res9: monocle.Iso[Bar.type,Unit] = monocle.PIso$$anon$10@5520ac34
   * }}}
   *
   * Finally, `GenIso.fields` is a whitebox macro which generalise `GenIso.apply` to all case classes:
   */
  def exerciseGenIsoFields(res0: (String, Int)) =
    GenIso.fields[Person].get(Person("John", 42)) should be(res0)

  /**
   * Be aware that whitebox macros are not supported by all IDEs.
   *
   * == Laws ==
   *
   * An `Iso` must satisfy all properties defined in `IsoLaws` from the core module. You can check the validity of your own `Iso` using `IsoTests` from the law module.
   *
   * In particular, an Iso must verify that `get` and `reverseGet` are inverse. This is done via `roundTripOneWay` and `roundTripOtherWay` laws:
   */
  def exerciseLaws(res0: Boolean, res1: Boolean) = {
    personToTuple.get(Person("Zoe", 25))
    def roundTripOneWay[S, A](i: Iso[S, A], s: S): Boolean =
      i.reverseGet(i.get(s)) == s

    def roundTripOtherWay[S, A](i: Iso[S, A], a: A): Boolean =
      i.get(i.reverseGet(a)) == a

    roundTripOneWay(personToTuple, Person("Zoey", 25)) should be(res0)

    roundTripOtherWay(personToTuple, ("Zoe", 52)) should be(res1)
  }

}


src/main/scala/monocle/LensExercises.scala
/*
 * Copyright 2017-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monoclelib

import monocle.Lens
import monocle.macros.GenLens
import monocle.macros.Lenses
import org.scalaexercises.definitions._
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

object LensHelper {
  case class Address(streetNumber: Int, streetName: String)

  val address      = Address(10, "High Street")
  val streetNumber = GenLens[Address](_.streetNumber)

  def neighbors(n: Int): List[Int] =
    if (n > 0) List(n - 1, n + 1) else List(n + 1)

  case class Person(name: String, age: Int, address: Address)
  val john = Person("John", 20, address)

  val addressLens = GenLens[Person](_.address)

  @Lenses
  case class Point(x: Int, y: Int)
  val p = Point(5, 3)

  @Lenses("_")
  case class OtherPoint(x: Int, y: Int)
  val op = OtherPoint(5, 3)

}

/**
 * == Lens ==
 *
 * A [[http://julien-truffaut.github.io/Monocle/optics/lens.html Lens]] is an optic used to zoom inside a `Product`, e.g. `case class`, `Tuple`, `HList` or even `Map`.
 *
 * `Lenses` have two type parameters generally called `S` and `A`: `Lens[S, A]` where `S` represents the `Product` and `A` an element inside of `S`.
 *
 * Let’s take a simple case class with two fields:
 * {{{
 *   case class Address(strNumber: Int, streetName: String)
 * }}}
 *
 * We can create a `Lens[Address, Int]` which zooms from an `Address` to its field `strNumber` by supplying a pair of functions:
 *
 *  - `get: Address => Int`
 *  - `set: Int => Address => Address`
 *
 * {{{
 *   import monocle.Lens
 *   val strNumber = Lens[Address, Int](_.strNumber)(n => a => a.copy(strNumber = n))
 * }}}
 *
 * This case is really straightforward so we automated the generation of `Lenses` from case classes using a macro:
 * {{{
 *   import monocle.macros.GenLens
 *   val strNumber = GenLens[Address](_.strNumber)
 * }}}
 *
 * @param name lens
 */
object LensExercises extends AnyFlatSpec with Matchers with Section {

  import LensHelper._

  /**
   * Once we have a `Lens`, we can use the supplied get and set functions (nothing fancy!):
   * {{{
   *   val address = Address(10, "High Street")
   * }}}
   */
  def exerciseGetAndSet(res0: Int, res1: Address) = {

    streetNumber.get(address) should be(res0)
    streetNumber.set(5)(address) should be(res1)
  }

  /**
   * We can also `modify` the target of `Lens` with a function, this is equivalent to call `get` and then `set`:
   */
  def exerciseModify(res0: Address, res1: Int, res2: Address) = {
    streetNumber.modify(_ + 1)(address) should be(res0)
    val n = streetNumber.get(address)
    n should be(res1)
    streetNumber.set(n + 1)(address) should be(res2)
  }

  /**
   * We can push the idea even further, with `modifyF` we can update the target of a `Lens` in a context, cf `cats.Functor`:
   * {{{
   *   def neighbors(n: Int): List[Int] =
   *   if(n > 0) List(n - 1, n + 1) else List(n + 1)
   *
   *   import cats.implicits._ // to get Functor[List] instance
   * }}}
   */
  def exerciseModifyF(res0: List[Address], res1: List[Address]) = {
    import cats.implicits._ // to get Functor[List] instance
    streetNumber.modifyF(neighbors)(address) should be(res0)
    streetNumber.modifyF(neighbors)(Address(135, "High Street")) should be(res1)
  }

  // TODO find a way to work with Future
  /**
   * This would work with any kind of `Functor` and is especially useful in conjunction with asynchronous APIs, where one has the task to update a deeply nested structure with the result of an asynchronous computation:
   *
   * {{{
   * import scala.concurrent._
   * import scala.concurrent.ExecutionContext.Implicits._ // to get global ExecutionContext
   *
   * def updateNumber(n: Int): Future[Int] = Future.successful(n + 1)
   * }}}
   * {{{
   *   strNumber.modifyF(updateNumber)(address)
   *   // res9: scala.concurrent.Future[Address] = Future(<not completed>)
   * }}}
   *
   * Most importantly, `Lenses` compose together allowing to zoom deeper in a data structure:
   *
   * {{{
   *     case class Person(name: String, age: Int, address: Address)
   *     val john = Person("John", 20, address)
   *
   *     val addressLens = GenLens[Person](_.address)
   * }}}
   */
  def exerciseComposeLens(res0: Int, res1: Person) = {

    (addressLens composeLens streetNumber).get(john) should be(res0)
    (addressLens composeLens streetNumber).set(2)(john) should be(res1)
  }

  /**
   * = Other Ways of Lens Composition =
   *
   * Is possible to compose few `Lenses` together by using `compose`:
   */
  def exerciseComposeMacro(res0: Person) = {
    val compose = GenLens[Person](_.name).set("Mike") compose GenLens[Person](_.age).modify(_ + 1)

    compose(john) shouldBe res0
  }

  /**
   * Same but with the simplified macro based syntax:
   * {{{
   * import monocle.macros.syntax.lens._
   *
   * john.lens(_.name).set("Mike").lens(_.age).modify(_ + 1)
   * }}}
   *
   * (All `Setter` like optics offer `set` and `modify` methods that returns an `EndoFunction` (i.e. `S => S`) which means that we can compose modification using basic function composition.)
   *
   * Sometimes you need an easy way to update `Product` type inside `Sum` type - for that case you can compose `Prism` with `Lens` by using `some`:
   */
  def exerciseLensPrismComposition(res0: Option[Int]) = {
    import monocle.std.option.some
    import monocle.macros.GenLens

    case class B(c: Int)
    case class A(b: Option[B])

    val c = GenLens[B](_.c)
    val b = GenLens[A](_.b)

    (b composePrism some composeLens c).getOption(A(Some(B(1)))) shouldBe res0
  }

  /**
   * == Lens Generation ==
   *
   * `Lens` creation is rather boiler platy but we developed a few macros to generate them automatically. All macros are defined in a separate module (see [[http://julien-truffaut.github.io/Monocle/modules.html modules]]).
   * {{{
   *   import monocle.macros.GenLens
   *   val age = GenLens[Person](_.age)
   * }}}
   *
   * `GenLens` can also be used to generate `Lens` several level deep:
   */
  def exerciseLensGeneration(res0: Person) =
    GenLens[Person](_.address.streetName).set("Iffley Road")(john) should be(res0)

  /**
   * For those who want to push `Lenses` generation even further, we created `@Lenses` macro annotation which generate `Lenses` for all fields of a case class. The generated `Lenses` are in the companion object of the case class:
   *
   * {{{
   * import monocle.macros.Lenses
   *
   * @Lenses case class Point(x: Int, y: Int)
   * val p = Point(5, 3)
   * }}}
   */
  def exerciseLensMacroAnn(res0: Int, res1: Point) = {
    Point.x.get(p) shouldBe res0
    Point.y.set(0)(p) shouldBe res1
  }

  /**
   * You can also add a prefix to `@Lenses` in order to prefix the generated `Lenses`:
   * {{{
   * @Lenses("_") case class OtherPoint(x: Int, y: Int)
   * val op = OtherPoint(5, 3)
   * }}}
   */
  def exerciseLensMacroAnnPrefix(res0: Int) =
    OtherPoint._x.get(op) shouldBe res0

  /**
   * == Laws ==
   *
   * A `Lens` must satisfy all properties defined in `LensLaws` from the `core` module. You can check the validity of your own Lenses` using `LensTests` from the `law` module.
   *
   * In particular, a `Lens` must respect the `getSet` law which states that if you get a value `A` from `S` and `set` it back in, the result is an object identical to the original one. A side effect of this law is that `set` must only update the `A` it points to, for example it cannot increment a counter or modify another value.
   *
   * On the other hand, the `setGet` law states that if you `set` a `value`, you always `get` the same value back. This law guarantees that `set` is actually updating a value `A` inside of `S`.
   */
  def exerciseLaws(res0: Boolean, res1: Boolean) = {
    val streetNumber = Lens[Address, Int](_.streetNumber)(n => a => a.copy(streetNumber = n))

    def getSet[S, A](l: Lens[S, A], s: S): Boolean =
      l.set(l.get(s))(s) == s

    def setGet[S, A](l: Lens[S, A], s: S, a: A): Boolean =
      l.get(l.set(a)(s)) == a

    getSet(streetNumber, address) should be(res0)

    setGet(streetNumber, address, 20) should be(res1)
  }
}


project/ProjectPlugin.scala
import sbt.Keys._
import sbt._
import com.alejandrohdezma.sbt.github.SbtGithubPlugin

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = plugins.JvmPlugin && SbtGithubPlugin

  object autoImport {

    lazy val V = new {
      val cats: String                = "2.1.1"
      val monocle: String             = "2.0.4"
      val shapeless: String           = "2.3.3"
      val scala: String               = "2.13.2"
      val scalaExercises: String      = "0.6.0"
      val scalacheck: String          = "1.14.3"
      val scalacheckShapeless: String = "1.2.5"
      val scalatest: String           = "3.1.2"
      val scalatestplusScheck: String = "3.1.1.1"
    }

    def dep(artifactId: String) = "org.scala-exercises" %% artifactId % V.scalaExercises

    lazy val exercisesSettings = Seq(
      libraryDependencies ++= Seq(
        dep("exercise-compiler"),
        dep("definitions"),
        "org.typelevel"              %% "alleycats-core"            % V.cats,
        "org.typelevel"              %% "cats-core"                 % V.cats,
        "com.github.julien-truffaut" %% "monocle-core"              % V.monocle,
        "com.github.julien-truffaut" %% "monocle-macro"             % V.monocle,
        "com.chuusai"                %% "shapeless"                 % V.shapeless,
        "org.scalatest"              %% "scalatest"                 % V.scalatest,
        "org.scalacheck"             %% "scalacheck"                % V.scalacheck,
        "org.scalatestplus"          %% "scalacheck-1-14"           % V.scalatestplusScheck,
        "com.github.alexarchambault" %% "scalacheck-shapeless_1.14" % V.scalacheckShapeless
      )
    )
  }

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      organization := "org.scala-exercises",
      organizationName := "47 Degrees",
      organizationHomepage := Some(url("https://47deg.com")),
      scalaVersion := autoImport.V.scala,
      resolvers ++= Seq(
        Resolver.sonatypeRepo("snapshots"),
        Resolver.sonatypeRepo("releases")
      ),
      scalacOptions += "-Ymacro-annotations"
    )
}
